Cargo.toml

```diff
@@ -4,22 +4,25 @@ version = "0.1.0"
 edition = "2021"
 
 [dependencies]
-ark-ec = "0.4.2"
+ark-ec = "^0.4.0"
 ark-ff = "^0.4.0"
 ark-poly = "^0.4.0"
 ark-std = "^0.4.0"
 ark-crypto-primitives = { version = "^0.4.0", default-features = false, features = ["r1cs", "sponge"] }
 ark-relations = { version = "^0.4.0", default-features = false }
 ark-r1cs-std = { version = "^0.4.0", default-features = false }
+ark-circom = { git = "https://github.com/gakonst/ark-circom.git" }
+ark-bn254 = "0.4.0"
```

Comment:
One question, do you know if ark-circom / circom-compat can work with other curves (with arkworks generics) rather than just bn254? if so (ideally but I'm not sure if it's possible) we could avoid importing bn254 as dependency (but again I'm not sure if ark-circom supports other curves)

Reply:
Sorry for the late reply. It took me a while to look into it.

I think ark-circom only directly supports `bn254`, but I noticed that it is also possible with other elliptic curves. ark-circom takes `ark_ec::pairing::Pairing;` as generics in R1CS. The R1CS constraints below are the ones with 3 (A,B,C) sparse matrices and depend on `E(pairing)`. Since BN254 implemented the `pairing trait`, ark-circom runs with BN254.

If other elliptic curves also implement the pairing trait, then ark-circom can be run on other elliptic curves as well. Actually, `ark_ec::models::bls12::Bls12`, a dependency of `ark_bls12_381,` has a `pairing trait`, so it should be possible on other elliptic curves would also be possible.
```
pub struct R1CS<E: Pairing> {
    pub num_inputs: usize,
    pub num_aux: usize,
    pub num_variables: usize,
    pub constraints: Vec<Constraints<E>>,
    pub wire_mapping: Option<Vec<usize>>,
}
```
However, the main topic is whether the process of converting from circom to R1CS of folding-schemes can be done in abstract generics, omitting `ark-bn254`. The above R1CS constraints are defined by a `<E as Pairing>::ScalarField` for the value of the sparse matrix, and the `ScalarField` implements a `PrimeField trait`. 
This suggests that it is possible to output R1CS on `PrimeField` of folding-schemes from R1CS of circom, which is easier to handle from the `frontend trait` if unified with `PrimeField`.
```
pub type Constraints<E> = (ConstraintVec<E>, ConstraintVec<E>, ConstraintVec<E>);
pub type ConstraintVec<E> = Vec<(usize, <E as Pairing>::ScalarField)>;
```

I felt the mechanics of ark-works and rust were complicated to me, but this is what I understand now. I am planning to rewrite the implementation over the weekend.

Reply:
@arnaucube 
Finally, I could get converted R1CS for `folding_schemes` from `ark-circom` without the `BN254` curve. Actually, the `BN254` remains to be used in the test function, because the `ark-cricom`'s `Pairing` generic needs specific curves to try test functions. However, by removing the `BN254` dependencies in the main code, I think the `frontend trait` can easily deal with getting `folding_schemes`'s `R1CS` which is on `PrimeField` rather than `BN254`.

I would be very happy to review these codes again. If possible, I would like to work on the design of the `frontend trait` next. Sorry for the back and forth on the order of issues.

Thank you.

Reply:
Oh that's very cool!!
Yeah for the tests is fine, we also use specific curves for some tests in other directories only for tests.

Maybe the dependency now can be moved into Cargo.toml's line 26 under `[dev-dependencies]`?

Reply:
I moved `ark-bn254` into `[dev-dependencies]`.